### PR TITLE
Have a single ActivateVolumeOptions struct for the ActivateVolumeWith* family

### PR DESCRIPTION
**Pull request by pedronis**
Also rename PIN in its definition to Passphrase to move toward
better and also more abstract naming.

Same in the corresponding RecoveryKeyUsageReason code.

See also issue #113

### CODE REVIEW

File: crypt.go [hunk 1]
```diff
@@ -383,20 +384,20 @@ func activateWithTPMKey(tpm *TPMConnection, volumeName, sourceDevicePath, keyPat
...
 		var authPrivateKey []byte
 
-		for ; pinTries > 0; pinTries-- {
+		for ; passphraseTries > 0; passphraseTries-- {
 			var pin string
 			if k.AuthMode2F() == AuthModePIN {
-				r := pinReader
-				pinReader = nil
+				r := passphraseReader
+				passphraseReader = nil
```
Comment (bboozzoo) [inline crypt.go:400]: Unrelated to renames, is this a precaution that the reader wouldn't be used again by accident?

File: crypt.go [hunk 2]
```diff
@@ -454,40 +455,55 @@ func makeActivateOptions(in []string) ([]string, error) {
...
 	KeyringPrefix string
 }
 
 // ActivateVolumeWithTPMSealedKey attempts to activate the LUKS encrypted volume at sourceDevicePath and create a mapping with the
 // name volumeName, using the TPM sealed key object at the specified keyPath. This makes use of systemd-cryptsetup.
 //
-// If the TPM sealed key object has a PIN defined, then this function will use systemd-ask-password to request it. If pinReader is not
-// nil, then an attempt to read the PIN from this will be made instead by reading all characters until the first newline. The PINTries
-// field of options defines how many attempts should be made to obtain the correct PIN before failing.
+// If the TPM sealed key object has a user passphrase/PIN defined, then this function will use systemd-ask-password to request it. If passphraseReader is not
```
Comment (bboozzoo) [inline crypt.go:500]: That's a long comment, but I see it's consistent with the rest of the comments in the package. Nevertheless, it does not fit into a file view in a github review page, making it a bit hard to follow. Could we unify the comments style with snapd at some point?

File: crypt.go [hunk 3]
```diff
@@ -454,40 +455,55 @@ func makeActivateOptions(in []string) ([]string, error) {
...
+	// UserPassphraseTries specifies the maximum number of times
+	// that unsealing with a user passphrase should be attempted
+	// before failing with an error and falling back to activating
+	// with the recovery key if RecoveryKeyTries is greater than
+	// zero. Setting this to zero disables unsealing with a user
+	// passphrase - in this case, an error will be returned if the
+	// sealed key object indicates that a user passphrase has been
+	// set. With a TPM attempts to unseal with a user passphrase
+	// will stop if the TPM enters dictionary attack lockout mode
+	// before this limit is reached.
```
Comment (bboozzoo) [inline crypt.go:470]:
```suggestion
	// set. With a TPM, subsequent attempts to unseal with a user passphrase
	// will be ceased if the TPM enters dictionary attack lockout mode
	// before the tries limit is reached.
```

File: crypt.go [hunk 4]
```diff
@@ -454,40 +455,55 @@ func makeActivateOptions(in []string) ([]string, error) {
...
+	// It is unused by ActivateWithRecoveryKey.
+	UserPassphraseTries int
```
Comment (bboozzoo) [inline crypt.go:472]: UserPassphraseMaxTries?

File: crypt.go [hunk 5]
```diff
@@ -454,40 +455,55 @@ func makeActivateOptions(in []string) ([]string, error) {
...
+
+	// RecoveryKeyTries specifies the maximum number of times that
+	// activation with the fallback recovery key should be
+	// attempted.  It is considered directly by
+	// ActivateWithRecoveryKey and when falling in the other
+	// cases, for example failed TPM unsealing.  Setting this to
+	// zero will disable attempts to activate with the fallback
+	// recovery key.
 	RecoveryKeyTries int
```
Comment (bboozzoo) [inline crypt.go:481]: RecoveryKeyMaxTries?

File: crypt.go [hunk 6]
```diff
@@ -454,40 +455,55 @@ func makeActivateOptions(in []string) ([]string, error) {
...
-	// if activation with the TPM sealed key fails, before failing with an error. Setting this to zero will disable attempts to activate
-	// with the fallback recovery key.
+// ActivateVolumeOptions provides options to the ActivateVolumeWith*
+// family of functions.
+type ActivateVolumeOptions struct {
+	// UserPassphraseTries specifies the maximum number of times
+	// that unsealing with a user passphrase should be attempted
+	// before failing with an error and falling back to activating
+	// with the recovery key if RecoveryKeyTries is greater than
+	// zero. Setting this to zero disables unsealing with a user
```
Comment (bboozzoo) [inline crypt.go:465]:
```suggestion
	// with the recovery key (see RecoveryKeyTries).
	// Setting this to zero disables unsealing with a user
```

File: crypt.go [hunk 7]
```diff
@@ -454,40 +455,55 @@ func makeActivateOptions(in []string) ([]string, error) {
...
 
-	// ActivateOptions provides a mechanism to pass additional options to systemd-cryptsetup.
+	// ActivateOptions provides a mechanism to pass additional
+	// options to systemd-cryptsetup.
 	ActivateOptions []string
 
-	// LockSealedKeyAccess controls whether LockAccessToSealedKeys should be called after unsealing the TPM sealed key. It is called if
-	// this is set to true, and not called if this is set to false.
-	LockSealedKeyAccess bool
+	// LockSealedKeys controls whether access to the keys locked
```
Comment (bboozzoo) [inline crypt.go:487]:
```suggestion
	// LockSealedKeys controls whether access to the keys is locked
```

**Review by bboozzoo — APPROVED** [review]
LGTM. Some comment tweaks for further consideration.

File: crypt.go [hunk 8]
```diff
@@ -454,40 +455,55 @@ func makeActivateOptions(in []string) ([]string, error) {
...
+	// after unsealing. That means calling LockAccessToSealedKeys
+	// in TPM case.
```
Comment (anonymouse64) [inline crypt.go:489]:
```suggestion
	// after unsealing by calling LockAccessToSealedKeys in the TPM case 
	// for example
```

**Review by anonymouse64 — APPROVED** [review]
some suggestions/wonderings that are pretty nitpicky /o\, but otherwise lgtm

File: crypt.go (hunk 1 above)
Comment (pedronis) [reply crypt.go:400]: I'm not quite sure what was the original intention here. It's meant to fall back to use systemd-ask-password after consuming one passphrase.  But it's very specific behavior that's another reason to have some more control later via the Prompter if needed.
